pkgs/apps/q-chem/installer.nix

```diff
@@ -0,0 +1,132 @@
+{ stdenv, lib, requireFile, autoPatchelfHook, tcsh, perl, openssh
+, writeScriptBin, runtimeShell, mpi
+, minorVersion ? 4
+, scratchDir ? "/tmp"
+, memory ? 2000
+} :
+
+assert
+  lib.asserts.assertMsg
+  (builtins.elem minorVersion [ 1 2 3 4 ])
+  "Unsupported version of Q-Chem";
+
+let
+  url = "https://www.q-chem.com/install/#linux";
+  qchemSrc-5_1 = requireFile {
+    inherit url;
+    name = "qc5${builtins.toString minorVersion}.tar";
+    sha256 = "71b0d7fd4f6b47a090e25267c53fb38f5dbc50a8d159f2e8669ecba8f88f5d96";
+  };
+  qchemSrc-5_2 = requireFile {
+    inherit url;
+    name = "qc5${builtins.toString minorVersion}.tar";
+    sha256 = "cf513b8215369d9e904f7f22bca5f8d43a39b3fd166c2e4ddf5f67965b50a6fa";
+  };
+  qchemSrc-5_3 = requireFile {
+    inherit url;
+    name = "qc5${builtins.toString minorVersion}.tar";
+    sha256 = "5e616542e3bd20ef299fed97ef8f8ab30c3fa264cdc0c8fe900bfa30d975c3d8";
+  };
+  qchemSrc-5_4 = requireFile {
+    inherit url;
+    name = "qc5${builtins.toString minorVersion}.tar";
+    sha256 = "5f62677c17fc62f6da2eb6d3f43efcc9569e5feea9313c870729c562aa3e41a9";
+  };
+
+  # Runs the installer script and prepares valid installation in the store.
+  # License will be missing after the installation, thus this preliminary
+  # version will not work, yet. Use the activation script to obtain a license.
+  qchemInit = stdenv.mkDerivation rec {
+    pname = "Q-Chem";
+    version = "5.${builtins.toString minorVersion}";
+
+    src =
+      if minorVersion == 1 then qchemSrc-5_1
+      else if minorVersion == 2 then qchemSrc-5_2
+      else if minorVersion == 3 then qchemSrc-5_3
+      else if minorVersion == 4 then qchemSrc-5_4
+      else "";
+
+    nativeBuildInputs = [ autoPatchelfHook ];
+    buildInputs = [ stdenv.cc.cc.lib ];
+    runtimeDependencies = buildInputs;
+    propagatedBuildInputs = [ tcsh perl openssh ];
+
+    sourceRoot = if minorVersion == 4 then "qc54_distrib" else ".";
+
+    postPatch = ''
+      patchShebangs ./qcinstall.sh
+    '';
+
+    dontConfigure = true;
+    dontBuild = true;
+
+    # Run the interactive configuration script.
+    # It will prepare a file to be sent to license@q-chem.com,
+    # but the HostIds will be missing, as get_hostid has not been fixed yet.
+    installPhase = ''
+      ./qcinstall.sh << EOF
```

Comment:
I am not entirely sure how exactly your alternative idea would look like. Basically replacing/patching `qcinstall` with our own nixified script, execute it interactively, imperatively, and with network connection and installing somewhere, e.g. `/opt/q-chem` and then wrapping as with Gaussian?

Certainly, the installer complicates things and the symlink stuff to avoid a second copy in the store is also *interesting*. But it has some advantages over the above idea (where I am not certain that you mean exactly this);

  - it is functional and involves no imperative step, that somehow influences anything besides the license.
  - If I should add the MPI versions later, the derivation is aware of the MPI requirement and make the appropriate settings from the context of the derivation
  - external installations in `/opt` or whatever require the sandbox path settings and those are only accessible for an admin. Thus, the Q-Chem installation would be exclusive to admins and normal users could not use the Q-Chem derivation.

The last point would be crucial for us, but as I said; maybe I got you wrong.

Reply:
Also, I don't understand `qcinstall.sh` completely. It's a 750 lines bash script, which then again also does not everything, as far as I understand it. The complete installation procedure creates more files and directories, than what I see in the script.

Reply:
@markuskowa as by your suggestions I also rewrote the activation script to use plain ssh instead of MPI. However, there is a problem, that we didn't think about. If of course requires and SSH server running. In my case, where I just want to activate my local workstation and no SSH server is running, simply dialing in via SSH doesn't work of course. MPI then again, works, as it figures out that for localhost SSH is not required. I tried to circumvent it now with a small bash if.